src/circuit/gadget/sinsemilla/note_commit.rs

```diff
@@ -0,0 +1,1721 @@
+                .chain(Some(("a_prime_check", a_prime_check)))
+                .chain(Some(("b3_c_prime_check", b3_c_prime_check)))
+                .chain(Some(("e1_f_prime_check", e1_f_prime_check)))
+                .chain(Some(("g1_g2_prime_check", g1_g2_prime_check)))
+                .chain(Some(("b_check", b_check)))
+                .chain(Some(("d_check", d_check)))
+                .chain(Some(("e_check", e_check)))
+                .chain(Some(("g_check", g_check)))
+                .chain(Some(("h_check", h_check)))
+                .chain(Some(("gd_x_check", gd_x_check)))
+                .chain(Some(("pkd_x_check", pkd_x_check)))
+                .chain(Some(("value_check", value_check)))
+                .map(move |(name, poly)| (name, q_canon_1.clone() * poly))
+        });
+
+        meta.create_gate("Canonicity checks", |meta| {
+            /*
+                a (250 bits) = bits 0..=249 of x(g_d)
+                b (10 bits)  = b_0 || b_1 || b_2 || b_3
+                            = (bits 250..=253 of x(g_d)) || (bit 254 of x(g_d)) || (ỹ bit of g_d) || (bits 0..=3 of pk★_d)
+                c (250 bits) = bits 4..=253 of pk★_d
+                d (60 bits)  = d_0 || d_1 || d_2 || d_3
+                            = (bit 254 of x(pk_d)) || (ỹ bit of pk_d) || (0..=7 of v) || (8..=57 of v)
+                e (10 bits)  = e_0 || e_1
+                            = (bits 58..=63 of v) || (bits 0..=3 of rho)
+                f (250 bits) = bits 4..=253 inclusive of rho
+                g (250 bits) = g_0 || g_1 || g_2
+                            = (bit 254 of rho) || (bits 0..=8 of psi) || (bits 9..=248 of psi)
+                h (10 bits)  = h_0 || h_1 || h_2
+                            = (bits 249..=253 of psi) || (bit 254 of psi) || 4 zero bits
+
+                |         A_0         |          A_1           |         A_2            |          A_3            |  A_4  |  A_5  |  A_6  |  A_7   |  A_8  |   A_9   |  q_canon_1  |  q_canon_2  |
+                ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+                |       a_prime       |       b3_c_prime       |       e1_f_prime       |       g1_g2_prime       |   a   |   b   |  b_2  |   b_3  |   c   |    d    |      0      |      0      |
+                |         d_1         |           d_2          |          z1_d          |           e             |  e_0  |  e_1  |   f   |    g   |  g_1  |    h    |      1      |      0      |
+                |         h_0         |           h_1          |         x(g_d)         |        x(pk_d)          | value |  b_0  |  b_1  |   d_0  |  g_0  |   z1_g  |      0      |      1      |
+                |a_prime_decomposition|b3_c_prime_decomposition|e1_f_prime_decomposition|g1_g2_prime_decomposition| z13_a | z13_c | z13_f | z13_g  |  psi  |   rho   |      0      |      0      |
```

Comment:
The simplest change to remove those rotations would be:
```suggestion
                |         A_0         |          A_1           |         A_2            |          A_3            |  A_4  |  A_5  |  A_6  |  A_7   |  A_8  |   A_9   |  q_canon_1  |  q_canon_2  |
                ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
                |       a_prime       |       b3_c_prime       |       e1_f_prime       |       g1_g2_prime       |   a   |   b   |  b_2  |   b_3  |   c   |    d    |      0      |      0      |
                |         d_1         |           d_2          |          z1_d          |           e             |  e_0  |x(g_d) |x(pk_d)|    g   | value |    h    |      1      |      0      |
                |         h_0         |           h_1          |          e_1           |           f             |  g_1  |  b_0  |  b_1  |   d_0  |  g_0  |   z1_g  |      0      |      1      |
                |a_prime_decomposition|b3_c_prime_decomposition|e1_f_prime_decomposition|g1_g2_prime_decomposition| z13_a | z13_c | z13_f | z13_g  |  psi  |   rho   |      0      |      0      |
```
(with equivalent changes to the first gate).

I'll have a think about whether I'd prefer an alternate region layout that more clearly conveys the data dependencies, but this change would suffice for now.

Reply:
I've figured out an alternate region layout that keeps almost all cells close to the ones they are involved in (which would make visualising constraints easier), while also keeping the above rotation reduction:
```suggestion
                |   A_0    |    A_1    |     A_2      | A_3 |  A_4  |       A_5     |    A_6   |     A_7      |  A_8  |    A_9    |  q_canon_1  |  q_canon_2  |
                -----------------------------------------------------------------------------------------------------------------------------------------------
                |    b     |     d     |      e       |  g  |   h   |       d_1     |  x(pk_d) |     b_3      |a_prime|    b_2    |      0      |      0      |
                |e1_f_prime|g1_g2_prime|    value     | d_2 |  z1_d |       e_0     |b3_c_prime|      c       |   a   |  x(g_d)   |      1      |      0      |
                |   e_1    |     f     |     g_0      | g_1 |  z1_g |       h_0     |    h_1   |     d_0      |  b_0  |    b_1    |      0      |      1      |
                |   rho    |   z13_f   |z14_e1_f_prime| psi | z13_g |z14_g1_g2_prime|  z13_c   |z14_b3_c_prime| z13_a |z13_a_prime|      0      |      0      |
```

Reply:
Small change needed to this comment: `z14_g1_g2_prime` -> `z13_g1_g2_prime`.

Otherwise, this has been implemented at  7aa3174880cb7fe26b1e594057675b4d7ff49087 and documented at ac5404a94317623ae713d6f0408d37247d46ea10.